### PR TITLE
use StorageDoubleMap iter keys

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: modules/margin-protocol/src/lib.rs [hunk 1]
```diff
@@ -339,10 +330,10 @@ impl<T: Trait> Module<T> {
 
 	fn _close_position(who: &T::AccountId, position_id: PositionId, price: Option<Price>) -> DispatchResult {
 		let position = Self::positions(position_id).ok_or(Error::<T>::PositionNotFound)?;
-		let index = Self::positions_by_trader(who, position.pool)
-			.iter()
-			.position(|id| *id == position_id)
-			.ok_or(Error::<T>::PositionNotOpenedByTrader)?;
+		ensure!(
+			Self::positions_by_trader(who, (position.pool, position_id)).is_some(),
```
Comment (xlc) [inline modules/margin-protocol/src/lib.rs:334]: In general, should use `exists` to check if a value exists so it does not need to decode the value.
In this case, the value is `()` so there are no overhead.

File: modules/margin-protocol/src/lib.rs [hunk 2]
```diff
@@ -503,11 +490,27 @@ impl<T: Trait> Module<T> {
...
+		<PositionsByTrader<T>>::insert(who, (pool, id), ());
+		PositionsByPool::insert(pool, (pair, id), ());
 
 		Ok(())
 	}
+
+	fn _ensure_can_open_more_position(who: &T::AccountId, pool: LiquidityPoolId, pair: TradingPair) -> DispatchResult {
+		let count = PositionsByPool::iter(pool).filter(|((p, _), _)| *p == pair).count();
+		ensure!(
+			count < T::GetPoolMaxOpenPositions::get(),
```
Comment (xlc) [inline modules/margin-protocol/src/lib.rs:502]: Use `<=`

File: modules/margin-protocol/src/lib.rs [hunk 3]
```diff
@@ -503,11 +490,27 @@ impl<T: Trait> Module<T> {
...
+			Error::<T>::CannotOpenMorePosition
+		);
+		let count = <PositionsByTrader<T>>::iter(who)
+			.filter(|((p, _), _)| *p == pool)
+			.count();
+		ensure!(
+			count < T::GetTraderMaxOpenPositions::get(),
```
Comment (xlc) [inline modules/margin-protocol/src/lib.rs:509]: Use `<=`

File: modules/margin-protocol/src/lib.rs [hunk 4]
```diff
@@ -503,11 +490,27 @@ impl<T: Trait> Module<T> {
...
-		<PositionsByTrader<T>>::mutate(who, pool, |ids| ids.push(id));
-		PositionsByPool::mutate(pool, pair, |ids| ids.push(id));
+		<PositionsByTrader<T>>::insert(who, (pool, id), ());
+		PositionsByPool::insert(pool, (pair, id), ());
 
 		Ok(())
 	}
+
+	fn _ensure_can_open_more_position(who: &T::AccountId, pool: LiquidityPoolId, pair: TradingPair) -> DispatchResult {
+		let count = PositionsByPool::iter(pool).filter(|((p, _), _)| *p == pair).count();
```
Comment (xlc) [inline modules/margin-protocol/src/lib.rs:500]: I am sure there are some low level magic to do counting without iterate everything...

Something to be revisited later.

Comment (xlc) [reply modules/margin-protocol/src/lib.rs:500]: https://github.com/laminar-protocol/laminar-chain/issues/153

File: modules/margin-protocol/src/lib.rs (hunk 2 above)
Comment (ermalkaleci) [reply modules/margin-protocol/src/lib.rs:502]: @xlc <= will be max+1

Comment (xlc) [reply modules/margin-protocol/src/lib.rs:502]: Oh I missed that check before add, not after

**Pull request merged**